### DISCUSSION TITLE
Alternative building app idea.

### DISCUSSION THREAD

**Discussion by M4rkY4** [airsdk/Adobe-Runtime-Support#2164] [Ideas]
Hi,

seeing the promising plans of Harman AIR devs in ActionScript evolving and the amount of work they plan to complete, I would like to address another problem (especially for newcomers) and an idea how to solve it.

One of the most complicated things on AIR right now is setting up builds (especially for mobile).
Setting up XML right is quite challenging, setting up Ant builds even more.
As far as I know, APM is meant to solve some of that problems, I've never used it, but it seems to be still a terminal based experience.

I love how "publishing" interface in Adobe Flash Pro worked, but now it's way too outdated (at least, for mobile) and is inside Adobe app.

What would you think of an idea of creating an alternative to that AIR app that auto-setups and initiates builds creations?

So that to make a build, a developer opens an app, click on checkboxes or select images and click "publish"? And that app runs APM, Ant or whatever needed, without showing terminal for the public (at least if not requested by a user)?

If there's a problem in design, Harman devs can create an open source project and I can contribute first few versions of launchable .FLA file with a design and buttons handling code for that.

As for me, exporting a project seems to become a bottleneck to AIR and it would be amazing if it can be solved.

What do you think?

**Comment by ajwfrost** [comment]
Hi

I completely agree, there's still a bit of a challenge to setting up and starting out with the SDK and with projects/apps. For what you're suggesting here, I think it's mostly a case of putting a user interface on top of the `apm` command-line application?

@marchbold  I believe the GUI idea was already suggested for a "v2" of `apm`, wondering if there's an option to have other folk contribute towards this goal?

Re-reading the suggestion here, we've then also got the "building" part of this to consider. `apm` is for setting up the application descriptor and pulling in the dependencies, but then the idea of publishing something is also needed. I tend to mostly use the command line (or set up ant for anything that needs to be scripted) and do find myself repeatedly typing the same things.. so maybe this could be added as an extension (or added in the GUI version of `apm` but then using additional tools to do stuff, i.e. the GUI would be more than just a package manager).. -> npm has "npm start" etc so it would be like that? So you add e.g. your iOS provisioning profile and the certificates you want to use, list out the files/folders to include, ANE/resource directories, etc... and then can just hit "build" or "publish for Android" etc..

Nice idea :-)

**Reply by M4rkY4** [reply]
I'm thinking more of a global approach, like with Tabs with UI element for each entry:

- Global tab =>
-- new / save / load project
-- AIR SDK link
-- JDK link
-- select assets folder
-- generate IML file for Intellij IDEA
-- generate project setup file for VsCode
-- build WIn64/Mac/iOS/Android buttons that activate Ant files (runs invisible in background)
-- for windows => select CMD or Powershell where to execute all commands _(command length limit windows issue)_

- iOS tab => 
-- generate basic XML config
-- generate basic config file (PRODUCTION::false, CHEATS::false)
-- add path to certificate
-- update config with APM
-- generate custom resources ANE from selected files _(and auto-import it to build)_
-- generate Assets.car _(and auto-import it to build)_
-- generate icons from single 1024x1024 _(and auto-import them to build)_
-- generate basic launch storyboard (_and auto-import it to build)_
-- generate startup images _(they seem to be still needed for older iOS devices)_
-- generate Ant file

- Android tab =>
-- generate basic XML config
-- generate basic config file (PRODUCTION::false, CHEATS::false)
-- add path to certificate
-- update config with APM
-- generate google-services.json
-- generate custom resources ANE from selected files _(and auto-import it to build)_
-- generate icons from single 1024x1024 _(and auto-import them to build)_
-- import selected notifications sound and notifications icon
-- generate Ant file

- Ant tab =>
-- shows Ant build command
-- shows Ant logs

the same to Win64 and Mac
and if something new mandatory is added, add a new UI element to it

**Reply by marchbold** [reply]
Yeah we have been assessing ideas here currently, as there are so many different ways people currently build AIR applications, it's hard to decide what tool is going to be best. And this is why we originally built apm as a simple command line utility, as this can sit alongside any of the build processes currently. 

We do see several key additons going forward:
- a UI wrapper for apm (though where this sits I'm not quite sure yet); 
- expanding apm to include asset generation (including things like [Assets.car/launchscreen creation](https://github.com/distriqt/AIR-ImageScripts) and firebase config android resources);
- apm manage the [app descriptor completely](https://github.com/airsdk/apm/discussions/52), removing any manual xml editing;
- getting apm to setup an IDE, creating project files, adding libraries/extensions, [unpackaging anes for adl](https://github.com/airsdk/apm/discussions/106) etc;
- actually creating the [packaging command](https://github.com/airsdk/apm/discussions/159) or [mxmlc-config.xml](https://github.com/airsdk/apm/discussions/58) for an app and potentially executing it or constructing ant scripts etc;

While creating a separate packaging tool may be useful, devs still have to setup an IDE to do debugging etc so it's potentially double work for a developer, to keep their IDE in sync with the tool. 

So I had originally thought the IDE setup would be a good approach and started down this path but Animate is a major hurdle. Many devs seem to still use it as the publishing tool and there's just no simple way for `apm` to edit `fla`s and set all the `swc`/`ane` paths etc. So it's a difficult integration path. I suppose we could start investigating the `fla` format more, I do vaguely remember that it is now a zip `xfl` format which potentially gives us an integration path. But the question remains whether this is the best approach. 

Ideally I'd like to encourage people to use an IDE for development and packaging and only use Animate for asset generation. Then we could use `apm` to setup the `IDE`s, editing the iml module definitions for intellij or the asconfig file for vscode (and potentially the batch files for flash develop).

Then the last thought is maybe we should just start fleshing out the vscode plugin use that as the UI for apm alongside the asconfig file? 

So I suppose I see the current issues as:
- sync with IDEs to allow debugging easily;
- Animate integration;
- duplication of work / UI (creating confusion on where to configure / run things);

**Reply by M4rkY4** [reply]
Wow, that's great, that's almost everything I've asked :)

IDEA debug is way better than Animates, totally true.

The problem with building in Animate is a workflow issue - you just click few buttons and that's it, it's done.
If you deeply use Animate, that's very unclear and unintuitive how to remain your workflow without sacrificing your options (or if its possible at all), maybe a good tutorial on main page with link on Air SDK would help.

Maybe, we (flash community) can create an Animate -> IDEA / VsCode -> Debug workflow tutorial with screenshots if Animate/IDEA screenshots are OK on AIR main page. As an Animate fan, I can investigate and create my version of easy workflow tutorial in case I'll be able to avoid functionality, speed and ease-of-use sacrifices _(I'm unsure whether I will be able, though)_.

As for IDEA building, all mobile projects I've worked on had unresolvable problems (at least at that point, we'd never checked if something changed) on building AAB with IDE, so Ant seems to be as an only option for Android which eliminates an option of Android device debugging, so until AAB problem exists, so If you have only Android device, Animate Emulator with touch layer looks like a better way to go.

**Reply by marchbold** [reply]
@M4rkY4 Not really sure what you are talking about with the AAB issue on intellij. We have no problems using it for everything. It does require usage of the adt.cfg flag the same as Animate for the AAB but this works fine for us? AAB is not used for debugging anyways, it's a release format to send the app to google play. 

Personally I find everything about developing an app in animate painful :), adding an extension with dependencies can take a while, code completion / inspection and editing tools are practically useless, and config options are hidden just as much as they are in any other ide, potentially the ui for some of the app descriptor is better in animate, but generally if you are doing anything beyond the basics you need to manually edit it anyways. It's power definitely lies in the visual / animation side. 

Generally the flow I recommend (and that we use for all our apps requiring animate) is:
- create assets in animate;
- treat it as a visual / animation tool only, (ideally no as3 requiring maintenance);
- export as either:
  - symbols via a swc;
  - spritesheet via xml/png;
- switch to your ide and use these assets, either:
  - adding the swc as a library and importing the symbols as classes;
  - loading the spritesheet;

Some projects we have a series of swcs / spritesheets representing different parts / components of the app. 

This flow also allows designers to work on visual assets in animate independent of the code base for the app.

Potentially this is worth documenting and explaining more, but this has been our process since FlashBuilder was released many many years ago :)

**Reply by manishpatel0404** [reply]
One can also consider FlashDevelop for this. It has a simple workflow, has a built-in debugging mechanism that still works great. The code editor is a bit slow and clunky, but otherwise, FlashDevelop, although not currently under development, still does a fairly decent job, because it is highly customizable. For editing code, we use NotePad++ with a customized UDL for AS3.

**Reply by marchbold** [reply]
@manishpatel0404 FlashDevelop is windows only so not really a good solution, particularly with the majority of iOS development tools not being available on windows. I'd suggest if we did pick a tool it would have to be cross-platform, macos/windows as a minimum.

**Reply by M4rkY4** [reply]
@marchbold thanks for the answer.


The approach your team is using is great for apps and for simultaneous development with two or more devs, but (if I understand it correctly) it is not best for games.

It would be a slow workflow for developing a game where there's a single developer and a single animator, especially if the developer and the animator is the same person :)

And for a more serious game development, Starling Framework with either Animate -> Flump -> TexturePacker, or Spine would work way better, slower in development, but way more performant in graphics. Spine animation capabilities are way more powerful than those of Animate (and a single Spine price is cheeper than "renting" Animate).


There are few great technologies that are good for professional development with a team of developers and I would not consider Flash as a safe bet simply because now it's not popular enough to find a good Flash developer.

But if you are a single developer in your team, Flash has no alternatives in terms of speed of game development when you use Animate, nothing even close. And powerful Animate CC animation + layouting + interactivity tool is the core of that speed. Animation alone is way better in Spine. Layouting and interactivity way better in Unity. But all cleverly combined in a single package - nothing closely as good as Animate CC.

Animate Scene is my screen, one scene for a game, another for UI, everything is positioned inside Animate and linked from Animate Library to external files. I could give you links for my old flash games, if you would like to see about what combination of animations and interactivity I'm talking about. 

And that's what I love in Flash + Animate - way less time on routine, way more time on fun of creation. Flash shines the brightest in a single-developer Animate projects.


Concerning IDE, I like to take best from both worlds, layouting and animating everything in Animate, using labels and smallest frame scripts ( like stop(); or this.methodFromExternalFile("param"); ) in timeline and nothing more, but keeping all big code in multiple external files and using the lovely Intellij IDEA for editing that files. Animate custom components are a Unity-like experience and were also working great, though being buggy and nerve-wrecking to setup, I highly doubt their property values will be exported to SWC.

Debugging and building are two issues with this approach, using Animate eliminate the use of awesome IDEA debugger, so I believe if we can somehow find a way to combine Animate with IDEA/VsCode debugging and building and creating a good tutorial, most of flash developers would benefit from that.


But I've experimented and googled a bit, there's no simple way to move from Animate, main timeline from SWF/SWC is always imported empty into the project and if you want to continue using code linkings (Animate -> .as files), you should have two codebases, one /src that's compiled to SWC, another /src for IDEA and they should somehow conveniently communicate between each other. And while it's still "worth the price" for experienced Flash enthusiasts, it's not nearly intuitive for a single person who want to start learning Flash.

Concerning extensions, debug, code completion I fully support your point of view :)

**Reply by marchbold** [reply]
Did you ever use FlashBuilders "FLA project" (or whatever it was called)? This used FB as the editor and when you went to publish it would use Animate (Flash CC) to package your app but FB would attach to the debugger.  

Maybe @ajwfrost could ask Adobe if some hook like that was still available in Animate? If we could trigger a "publish" from the IDE but not get animate to attach then the IDE should be able to attach to whatever animate launches in debug mode. Might be an avenue for us to explore for those still in Animate?

Highly doubt it would ever be something I'd ever use, honestly even my solo projects I still follow the approach I've mentioned above, and yes tools like flump / texture packer etc feature in that flow. I feel once you have the process down development is quick, particularly if you create a solid theme with a series of components. But I do understand that the multiple avenues the Flash/AIR ecosystem have has been one of it's strengths in the past so I highly value other approaches :)

**Reply by M4rkY4** [reply]
No, I've never used FlashBuilder. I started coding inside Flash Professional, then FlashDevelop, then IDEA.

**Reply by ajwfrost** [reply]
It is possible to drive Animate behaviour externally, using the JSFL extensibility stuff... (maybe renamed JSAPI now?)
https://developer.adobe.com/animate/

**Reply by marchbold** [reply]
Right, so definitely an integration option then ... could be a good reason to start looking more seriously into building out the vscode functionality? Using vscode as the IDE alongside animate probably could work really well?
